Weighted poisson integration

jhenin opened the pull request:
Implements weighted Poisson integration of free energy gradients.
Also fixes a small error in Neumann boundary conditions for the existing unweighted case.

TODO:

- [x] revert formatting changes (braces and indents)
- [x] user flag to enable/disable
- [x] Documentation
- [x] Fix failing tests due to boundary condition fixes in unweighted Poisson
- [x] Add test for new feature, weighted Poisson

jhenin committed:
Weighted Poisson integration (squashed changes)

clean up code

try better initialisations

rename function with snake_case

clean code

implement laplacian matrix saving

implement ADAM optimizer

Split integrate_potential class to own files

levo-str committed:
cache the laplacian coefficients and integrate calculate weight sum in the laplacian loop to make the programme faster

levo-str committed:
precompute laplacian coefficient and parallelize code

levo-str committed:
implement data extrapolation

levo-str committed:
fix print of convergence

levo-str committed:
uses the conjugate gradient optimizer for unweighted poisson resolution

levo-str committed:
implement as minimization of error on data's estimation of the gradient

levo-str committed:
fix unweighted boundary conditions handling and change laplacian calculation when there is no precomputation of the coefficients

levo-str committed:
fix segmentation error in div calculation when handling periodic boundary condition and implement choice between on the fly laplacian computation or precompute coefficients

levo-str committed:
clean code

levo-str committed:
use optimal parameter

jhenin committed:
Don't include omp header

levo-str committed:
delete unused code

levo-str committed:
parallel for loop over signed int and force thresholds as floats

levo-str committed:
fix type incoherence

levo-str committed:
remove excess variable

levo-str committed:
reformat code as normal

levo-str committed:
[MUST BE CHECKED] add flag for weighted integration

levo-str committed:
comment code

jhenin committed:
chore: remove unused code in poisson_integrator

jhenin committed:
chore: remove unused code

jhenin committed:
chore: undo formatting changes

jhenin committed:
chore: more cleanup of weighted Poisson

jhenin committed:
fix: Adapt weighted Poisson for ABF

jhenin committed:
test: Update NAMD regression tests output

for small fix to BC in unweighted Poisson

levo-str committed:
allow eABF weighted Poisson integration

levo-str committed:
fix reading for eABF

jhenin committed:
fix: smart pointer to fix memory leak

jhenin committed:
test: add functional test for 2D eABF

levo-str committed:
:fix: fix pointer issue, computation size issue and divergence stencil calculation

levo-str committed:
fix the memory leakage by correcting border gradients calculation for potential extrapolation

levo-str committed:
fix the memory leakage by correcting border gradients calculation for potential extrapolation

levo-str committed:
add deletion of computation grid pointer

levo-str committed:
fix extrapolation (again)

levo-str committed:
delete unused logs

jhenin committed:
fix: add missing test inputs

levo-str committed:
update test and delete unused testing variable

levo-str committed:
add new test eabf test with weighted integration

levo-str committed:
add new test abf2d_euler_weighted.in

levo-str committed:
:fix: correct test's colvar file name

levo-str committed:
update tests [Needs to be checked]

jhenin committed:
fix: test for empty count grid before integrating

jhenin committed:
test: reduce grid size in ext abf2d test

levo-str committed:
add documentation

levo-str committed:
add missing argument

levo-str committed:
allow sample grid to be added later

levo-str committed:
add 2D eABF with weighted integration test

levo-str committed:
allow sample grid to be added later

levo-str committed:
format code

levo-str committed:
check if this passes the test

levo-str committed:
delete unused import

levo-str committed:
error catching to handle 1D cases

jhenin committed:
fix: more robust colvargrid_gradient constructor

jhenin committed:
fix:  grid handling in ABF, limit header scope

jhenin committed:
test: fix output for custom grid metadata in ABF

levo-str committed:
put back integrate weighted argument

levo-str committed:
improve test with custom grid of different size to the original one

levo-str committed:
update eABF weighted test reference value

levo-str committed:
delete solved comments and unused log

levo-str committed:
create ABF with weighted integration NAMD test

levo-str committed:
create NAMD test for eABF

levo-str committed:
update tests

jhenin committed:
Make total_forces_same_step consistent

in proxy_stubs wrt other proxys

jhenin committed:
Remove redundant gradient tests

HanatoK commented on namd/tests/library/018_pathCV/AutoDiff/test.pmf:
```diff
@@ -2,1220 +2,1220 @@
- -2.00000000000000e-01  1.40000000000000e+00   7.70437678236428e-02
- -2.00000000000000e-01  1.45000000000000e+00   7.69776442679124e-02
- -2.00000000000000e-01  1.50000000000000e+00   7.69557395004421e-02
+ -2.00000000000000e-01 -5.00000000000000e-01   5.66747557569237e-02
```
Why are the test data changed here? I know this PR fixes the boundary condition in the integrator but `018_pathCV` does not use any periodic CVs.

jhenin replied:
The fix applies to Neumann (non-periodic) boundary conditions.

HanatoK commented:
When restarting `000_eulerangles_ext_abf2d_euler_weighted`, NAMD crashes. The GDB backtrace:
```
#0  0x00007ffff763dd3c in __pthread_kill_implementation () from /lib64/libc.so.6
#1  0x00007ffff75e27b6 in raise () from /lib64/libc.so.6
#2  0x00007ffff75c934b in abort () from /lib64/libc.so.6
#3  0x00007ffff7889972 in std::__glibcxx_assert_fail(char const*, int, char const*, char const*) () from /lib64/libstdc++.so.6
#4  0x00000000009b1cd7 in std::vector<double, std::allocator<double> >::operator[] (this=0x1d85b08, __n=0) at /usr/include/c++/15/bits/stl_vector.h:1263
#5  0x00000000013ba889 in colvargrid_integrate::update_weighted_div_local (this=0x1d85770, ix0=std::vector of length 2, capacity 2 = {...}) at colvars/src/colvargrid_integrate.cpp:344
#6  0x00000000013b8029 in colvargrid_integrate::set_weighted_div (this=0x1d85770) at colvars/src/colvargrid_integrate.cpp:151
#7  0x00000000012b8869 in colvargrid_integrate::set_div (this=0x1d85770) at colvars/src/colvargrid_integrate.h:50
#8  0x00000000012c48fb in colvarbias_abf::read_state_data_template_<std::basic_istream<char, std::char_traits<char> > > (this=0x1d5a880, is=...) at colvars/src/colvarbias_abf.cpp:968
#9  0x00000000012b66ab in colvarbias_abf::read_state_data (this=0x1d5a880, is=...) at colvars/src/colvarbias_abf.cpp:1017
#10 0x00000000012a9a2a in colvarbias::read_state_template_<std::basic_istream<char, std::char_traits<char> > > (this=0x1d5a880, is=...) at colvars/src/colvarbias.cpp:641
#11 0x00000000012a43fb in colvarbias::read_state (this=0x1d5a880, is=...) at colvars/src/colvarbias.cpp:665
#12 0x00000000013d2c5e in colvarmodule::read_objects_state (this=0x1b92970, is=...) at colvars/src/colvarmodule.cpp:1811
#13 0x00000000013e7d2f in colvarmodule::read_state_template_<std::basic_istream<char, std::char_traits<char> > > (this=0x1b92970, is=...) at colvars/src/colvarmodule.cpp:1724
#14 0x00000000013d25b9 in colvarmodule::read_state (this=0x1b92970, is=...) at colvars/src/colvarmodule.cpp:1732
#15 0x00000000013d11fe in colvarmodule::setup_input (this=0x1b92970) at colvars/src/colvarmodule.cpp:1562
#16 0x00000000014274e0 in cvscript_cv_load (pobj=0x1b4b140, objc=3, objv=0x7fffffffb9b0) at colvars/src/colvarscript_commands.h:484
#17 0x000000000141c9c1 in colvarscript::run (this=0x1b4b140, objc=3, objv=0x7fffffffb9b0) at colvars/src/colvarscript.cpp:439
#18 0x000000000141e87f in tcl_run_colvarscript_command (my_interp=0x1a75410, objc=3, objv=0x1a79790) at colvars/src/colvarscript.cpp:740
#19 0x00007ffff7de3ab8 in TclNRRunCallbacks () from /lib64/libtcl8.6.so
#20 0x00007ffff7dea234 in ?? () from /lib64/libtcl8.6.so
#21 0x00007ffff7e90587 in Tcl_FSEvalFileEx () from /lib64/libtcl8.6.so
#22 0x00007ffff7e90807 in Tcl_EvalFile () from /lib64/libtcl8.6.so
#23 0x0000000001121e77 in ScriptTcl::load (this=0x1a742b0, scriptFile=0x7fffffffd459 "test.restart.namd") at src/ScriptTcl.C:2423
#24 0x00000000009c463c in after_backend_init (argc=2, argv=0x7fffffffce68) at src/mainfunc.C:181
#25 0x00000000009c3f88 in main (argc=3, argv=0x7fffffffce68) at src/mainfunc.C:52
```

jhenin commented on namd/tests/library/000_distance-grid_abf-customgrid/test.in:
```diff
@@ -31,6 +31,6 @@ abf {
         width 0.2
         lowerBoundary 1.0
-        upperBoundary 5.0
+        upperBoundary 6.0
```
This ensures that the custom grid has a different number of points than the default one based on the colvar's parameters. Because there was an accidental match in the previous test, some errors went unnoticed.

jhenin commented on misc_interfaces/stubs/colvarproxy_stub.cpp:
```diff
@@ -74,7 +74,7 @@ bool colvarproxy_stub::total_forces_enabled() const
 bool colvarproxy_stub::total_forces_same_step() const
 {
-  return total_force_requested;
+  return true;
```
I think so, but @giacomofiorin may have a different opinion. I didn't see a rationale for that flag changing based on total_force_requested, i.e. the fact that it's true does not indicate that total forces are enabled or requested or even available.

jhenin committed:
Improve poisson_integrator by adding CLI flags

uses CLI11.hpp, now we need to find a common place in the tree to share it with other tools

also fixes Doxygen comments are noted by Haochuan

jhenin commented on src/colvargrid_integrate.cpp:
```diff
@@ -382,313 +521,851 @@ void colvargrid_integrate::atimes(const std::vector<cvm::real> &A, std::vector<c
+         gradients->incr(ix)) {
+      size_t count = gradients->samples->value(ix);
+      if (count > 0) {
+        insert_into_sorted_list<size_t>(sorted_counts, count);
```
Agreed, thanks!

jhenin committed:
Cast integers to cvm::real before division

HanatoK commented on src/colvargrid_integrate.cpp:
```diff
@@ -146,92 +189,187 @@ void colvargrid_integrate::update_div_neighbors(const std::vector<int> &ix0)
+    cvm::real weight = regularized_weights[gradients->address(surrounding_point_coordinates)];
+
+    for (size_t i = 0; i < nd; i++) {
+      div_at_point += pow(-1, surrounding_point_relative_position[i] + 1) *
```
(-1)^n could be either -1 or 1 depending on whether n is an even number or an odd number, so this can be simply calculated without calling `pow`.

HanatoK commented on src/colvargrid_integrate.cpp:
```diff
@@ -146,92 +189,187 @@ void colvargrid_integrate::update_div_neighbors(const std::vector<int> &ix0)
+  surrounding_points_relative_positions.clear();
+  size_t n_combinations = pow(2, nd);
+  for (size_t i = 0; i < n_combinations; i++) {
+    surrounding_points_relative_positions.push_back(convert_base_two(i, nd));
```
I feel that the calls to `convert_base_two` is unnecessary because an integer has already all its bits itself. `convert_base_two` simply checks the bits of `i` and stores them into a vector of `nd` size. But maybe I am wrong, and I am not sure how to optimize this.

levo-str committed:
delete useless imports

levo-str committed:
add more details to the documentation

jhenin commented on src/colvargrid_integrate.h:
```diff
@@ -1,78 +1,195 @@
+  bool weighted = false;
+  bool precompute = true;
+
+  colvar_grid_scalar* computation_grid;
```
Thanks, since this is not managing any memory, we'll keep the raw pointer for simplicity but initialize to nullptr.

jhenin committed:
Cleanup after PR review

Thanks Haochuan for the feedback.

levo-str committed:
prevent div calculation before preparations

levo-str committed:
restore use import of sys/stat.h

levo-str committed:
optimize code thanks to HanatoK's recommendations

HanatoK committed:
fix: change the type of nb_averaged_gradients from size_t to int

Fix the compilation error of comparing signed and unsigned integers.

levo-str committed:
optimize code thanks to HanatoK's recommendations

jhenin committed:
Fixes after rebase

jhenin committed:
Re-indent two files with inconsistent spaces

levo-str committed:
fix wrong calculations of div "supplement" i.e. boundary handling

levo-str committed:
[fix] fix calculation of eABF gradients with count (for weighted integration)

levo-str committed:
rerun test

jhenin committed:
perf: Avoid calls to std::pow

jhenin commented on src/colvargrid_integrate.cpp:
```diff
@@ -146,92 +189,187 @@ void colvargrid_integrate::update_div_neighbors(const std::vector<int> &ix0)
+{
+  const size_t linear_index = computation_grid->address(ix0);
+  cvm::real div_at_point = 0;
+  for (std::vector<int> surrounding_point_relative_position :
```
Done.

jhenin committed:
opt: compute_averaged_border_normal_gradient

Reduce allocations by reserving containers and moving allocations
out of loops.
